File: src/main/java/hdzi/editstarters/buildsystem/ProjectDependency.java
package hdzi.editstarters.buildsystem;

import com.intellij.psi.PsiElement;
import hdzi.editstarters.dependency.Dependency;
import lombok.Data;

/**
 * Created by taojinhou on 2019/1/15.
 */
@Data
public class ProjectDependency implements Dependency {
    private final String groupId;
    private final String artifactId;
    private final String version;
    private final PsiElement element;

    public ProjectDependency(String groupId, String artifactId, String version, PsiElement element) {
        this.groupId = groupId;
        this.artifactId = artifactId;
        this.version = version;
        this.element = element;
    }

    public ProjectDependency(String groupId, String artifactId) {
        this(groupId, artifactId, null, null);
    }

    public ProjectDependency(String groupId, String artifactId, String version) {
        this(groupId, artifactId, version, null);
    }

    public ProjectDependency(String groupId, String artifactId, PsiElement element) {
        this(groupId, artifactId, null, element);
    }
}

File: src/main/kotlin/hdzi/editstarters/buildsystem/ProjectDependency.kt
package hdzi.editstarters.buildsystem

import com.intellij.psi.PsiElement
import hdzi.editstarters.springboot.Dependency

/**
 * Created by taojinhou on 2019/1/15.
 */
class ProjectDependency(
    override val groupId: String,
    override val artifactId: String,
    val version: String?
) : Dependency {
    constructor(groupId: String, artifactId: String) : this(groupId, artifactId, null)

    constructor(groupId: String, artifactId: String, element: PsiElement) : this(groupId, artifactId) {
        this.element = element
    }

    // 依赖的位置
    var element: PsiElement? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ProjectDependency

        if (groupId != other.groupId) return false
        if (artifactId != other.artifactId) return false

        return true
    }

    override fun hashCode(): Int {
        var result = groupId.hashCode()
        result = 31 * result + artifactId.hashCode()
        return result
    }
}